### PR TITLE
Add dining philosophers as an async exercise

### PR DESCRIPTION
**Pull request by rbehjati**
This PR adds Dining Philosophers as an async exercise. 
Since we already have a description of the dining philosophers problem, this PR mostly 
- Adds template for async dining philosophers, with a few brief notes about the use of `tokio`,
- Adds the async solution for dining philosophers,
- Updates the structure to show that now we have two Async Rust exercises.

@djmitche I did not remove the elevator exercise. I think we can keep it for now, until we find a simpler replacement for it. Do we have a complete solution for it anywhere? If so, we could add a link to it in the solution section.

@mgeisler I've added a redirect rule for the link I change, but I am not sure if it is right. Can you please check that?
- Update: Removed the redirect, as one of the CI steps seemed to fail because of it.

### CODE REVIEW

**Review by djmitche — APPROVED** [review]
This looks good -- very parallel to the first exercise, which should help students focus on the async bits. 

As to the elevator exercise: there is no solution, as the challenges are very open-ended. That was a mistake on my part! I don't think we should keep it -- it's too much "given" code, and too open-ended.  (to be clear: it's fine to leave it as-is in this PR, but we should still aim to replace it)

File: src/exercises/concurrency/dining-philosophers-async.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+use std::time::Duration;
+use tokio::time;
+use tokio::sync::mpsc::{self, Sender};
+use tokio::sync::Mutex;
+
+
+struct Fork;
+
+/// We've already learnt that to avoid a deadlock, we have to break the
+/// symmetry. So here we call the forks the first_fork and the second_fork.
```
[the remark this answers is not in the document]
Comment (rbehjati) [reply src/exercises/concurrency/dining-philosophers-async.md:25]: What do you mean by the description of the problem? Do you mean the description of the problem in the morning? 

Breaking the symmetry is stated in the solution. I think the idea is that you arrive at this solution or something similar to it during the exercise session. But perhaps this comment is confusing. I can remove it.

**Comment by rbehjati** [comment]
I am getting the following error when running `mdbook test`

```
---- /var/folders/4l/r2r03g8s1d933_tplz2_6d1800qfjw/T/mdbook-zPryzM/exercises/concurrency/solutions-afternoon.md - Concurrency_Afternoon_Exercise::Dining_Philosophers___Async (line 7) stdout ----
error[E0433]: failed to resolve: use of undeclared crate or module `tokio`
  --> /var/folders/4l/r2r03g8s1d933_tplz2_6d1800qfjw/T/mdbook-zPryzM/exercises/concurrency/solutions-afternoon.md:26:5
   |
20 | use tokio::sync::mpsc::{self, Sender};
   |     ^^^^^ use of undeclared crate or module `tokio`
```

How can I fix this?

**Comment by rbehjati** [comment]
> I believe the fix is to add `compile_fail` to the beginning of the codeblock.

Thanks. Looks like it is fixed now.

**Review by mgeisler — COMMENTED** [review]
Looks good!

File: src/exercises/concurrency/dining-philosophers-async.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+}
+```
+
+Since this time you are using Async Rust, you'll need a `tokio` dependency.
+You can use the following `Cargo.toml`:
+
+<!-- File Cargo.toml -->
+
+```toml
+[package]
```
Comment (mgeisler) [inline src/exercises/concurrency/dining-philosophers-async.md:41]: Would you not need this as a stand-alone `Cargo.toml` file committed to the repository? That way you can build the code using `cargo test` in CI.

**Pull request merged**